modules/reactor/src/meshgenerators/ExtraElementIDConnectionFixup.C

```diff
@@ -0,0 +1,231 @@
+
+    map_coarse_to_fine[id].insert(elem->id());
+    coarse_ids.insert(id);
+
+    for (auto & node : elem->node_ref_range())
+      map_node_to_fine[node.id()].insert(elem->id());
+  }
+
+  if (_verbose)
+    _console << " Number of coarse elements before fixup: " << coarse_ids.size() << std::endl;
+
+  while (!coarse_ids.empty())
+  {
+    const dof_id_type coarse_id = *coarse_ids.begin();
+    coarse_ids.erase(coarse_id);
+    const auto & fine_ids = map_coarse_to_fine[coarse_id];
+
+    // the coarse element could contain no fine element due to fixup
+    if (fine_ids.size() == 0)
+      continue;
+
+    // merge the coarse element with only a single fine element into its neighboring coarse element
+    if (fine_ids.size() == 1)
+    {
+      Elem * elem = mesh->elem_ptr(*fine_ids.begin());
+      for (auto neig : elem->neighbor_ptr_range())
+        if (neig)
+        {
+          const dof_id_type newid = neig->get_extra_integer(src_id);
+          if (_verbose)
+            _console << " Single element with ID " << elem->id() << ": " << coarse_id << " => "
+                     << newid << std::endl;
+          elem->set_extra_integer(src_id, newid);
+          map_coarse_to_fine[coarse_id].erase(elem->id());
+          map_coarse_to_fine[newid].insert(elem->id());
+
+          // put changed ids into re-consideration
+          coarse_ids.insert(newid);
+          break;
+        }
+
+      continue;
+    }
+
+    // fine element ids that has been visited
+    std::set<dof_id_type> visited;
+    // fine element buffer for visiting
+    std::vector<const Elem *> buffer;
+    // nodes of fine elements that have been visited
+    std::set<dof_id_type> visited_nodes;
+    // a copy of fine elements to track unvisited elements
+    auto fine_ids_copy = fine_ids;
+
+    while (!fine_ids_copy.empty())
+    {
+      buffer.push_back(mesh->elem_ptr(*fine_ids_copy.begin()));
+
+      // nodes of all elements in this current visit
+      std::set<dof_id_type> nodes;
+      while (buffer.size() != 0)
+      {
+        const Elem * elem = buffer.back();
+        visited.insert(elem->id());
+        fine_ids_copy.erase(elem->id());
+        buffer.pop_back();
+
+        for (unsigned int n = 0; n < elem->n_nodes(); ++n)
+          nodes.insert(elem->node_ptr(n)->id());
+
+        // push connected elements through sides with the same coarse ID into the buffer
+        for (auto neig : elem->neighbor_ptr_range())
+          if (neig && neig->get_extra_integer(src_id) == coarse_id &&
+              visited.count(neig->id()) == 0)
+            buffer.push_back(neig);
+      }
+
+      // merge nodes
+      if (!visited_nodes.empty())
+      {
+        std::set<dof_id_type> int_nodes;
+        std::set_intersection(visited_nodes.begin(),
+                              visited_nodes.end(),
+                              nodes.begin(),
+                              nodes.end(),
+                              std::inserter(int_nodes, int_nodes.begin()));
```

Comment:
what is this for?
add a comment

Reply:
Added: `        // find the nodes linking the visited regions and the current disconnected region`.